edx_notifications/channels/channel.py

```diff
@@ -0,0 +1,206 @@
+                if not mapping:
+                    search_name, __, __ = search_name.rpartition('.')
+
+            if not mapping:
+                # did we reach the end without a mapping found? Use the global wildcard '*'
+                mapping = mappings.get('*')
+                if not mapping:
+                    # oops, this shouldn't happen if things are configured correctly
+                    msg = (
+                        "NOTIFICATION_CHANNEL_PROVIDERS is not configured "
+                        "with a global '*' definition. This must always be defined."
+                    )
+                    raise ImproperlyConfigured(msg)
+
+        if mapping not in _CHANNEL_PROVIDERS:
+            msg = (
+                "Bad mapping of NotificationType to NotificationChannel. "
+                "Mapping was looking for NotificationChannel {name} but "
+                "it was not found".format(name=mapping)
+            )
+            raise ImproperlyConfigured(msg)
+
+        _CHANNEL_PROVIDERS_TYPE_MAPS[type_name] = _CHANNEL_PROVIDERS[mapping]
+
+    return _CHANNEL_PROVIDERS_TYPE_MAPS[type_name]
+
+
+def _get_channel_preference(user_id, msg_type):  # pylint: disable=unused-argument
+    """
+    Returns what the user has chosen to be his/her preference
+    for edx_notifications.
+
+    Return of None = no preference declared
+    """
+
+    # not yet implemented
+    return None
+
+
+def get_notification_channel(user_id, msg_type):
+    """
+    Returns the appropriate NotificationChannel
+    for this user and msg_type.
+
+    Ultimately, this will be user-selectable, e.g.
+    'send my discussion forum notifications to my mobile device via SMS'
+    but for now, we're always mapping to the system default
+
+    NOTE: When we switch over to gevent support, we should use some
+    locking techniques in this area to prevent concurrent execution
+    """
+
+    global _CHANNEL_PROVIDERS  # pylint: disable=global-statement, global-variable-not-assigned
+
+    if not _CHANNEL_PROVIDERS:
+        _init_channel_providers()
+
+    # first see what the user preference is
+    channel = _get_channel_preference(user_id, msg_type)  # pylint: disable=assignment-from-none
+    if not channel:
+        channel = _get_system_channel_mapping(msg_type.name)
+
+    return channel
+
+
+def reset_notification_channels():
+    """
+    Clear out all cached channel definitions and mappings.
+    This is useful for testing scenarious, but likely should not
+    be called in normal runtimes
+
+    NOTE: When we switch over to gevent support, we should use some
+    locking techniques in this area to prevent concurrent execution
+    """
+
+    global _CHANNEL_PROVIDERS  # pylint: disable=global-statement
+    global _CHANNEL_PROVIDERS_TYPE_MAPS  # pylint: disable=global-statement
+
+    _CHANNEL_PROVIDERS = {}
+    _CHANNEL_PROVIDERS_TYPE_MAPS = {}
+
+
+class BaseNotificationChannelProvider(object):
```

Comment:
This could perhaps be written shorter by just using a `namedtuple` as a baseclass.